daemon/k8s_watcher.go

```diff
@@ -741,11 +745,115 @@ func (d *Daemon) EnableK8sWatcher(queueSize uint) error {
 		cnpConverterFunc,
 		cnpStore,
 	)
-	d.blockWaitGroupToSyncResources(wait.NeverStop, ciliumV2Controller, k8sAPIGroupCiliumV2)
+	d.blockWaitGroupToSyncResources(wait.NeverStop, ciliumV2Controller, k8sAPIGroupCiliumNetworkPolicyV2)
 	go ciliumV2Controller.Run(wait.NeverStop)
-	d.k8sAPIGroups.addAPI(k8sAPIGroupCiliumV2)
+	d.k8sAPIGroups.addAPI(k8sAPIGroupCiliumNetworkPolicyV2)
 
 	asyncControllers := sync.WaitGroup{}
+	asyncControllers.Add(1)
+
+	// CiliumNode objects are used for node discovery until the key-value
+	// store is connected
+	go func() {
+		var once sync.Once
+		for {
+			_, ciliumNodeInformer := informer.NewInformer(
+				cache.NewListWatchFromClient(ciliumNPClient.CiliumV2().RESTClient(),
+					"ciliumnodes", v1.NamespaceAll, fields.Everything()),
+				&cilium_v2.CiliumNode{},
+				0,
+				cache.ResourceEventHandlerFuncs{
+					AddFunc: func(obj interface{}) {
+						var valid, equal bool
+						defer func() { d.K8sEventReceived(metricCiliumNode, metricCreate, valid, equal) }()
+						if ciliumNode, ok := obj.(*cilium_v2.CiliumNode); ok {
+							valid = true
+							n := node.ParseCiliumNode(ciliumNode)
+							if n.IsLocal() {
+								equal = true
+								return
+							}
+							serNodes.Enqueue(func() error {
+								d.nodeDiscovery.Manager.NodeUpdated(n)
+								d.K8sEventProcessed(metricCiliumNode, metricCreate, true)
+								return nil
+							}, serializer.NoRetry)
+						}
+					},
+					UpdateFunc: func(oldObj, newObj interface{}) {
+						var valid, equal bool
+						defer func() { d.K8sEventReceived(metricCiliumNode, metricUpdate, valid, equal) }()
+						if ciliumNode, ok := newObj.(*cilium_v2.CiliumNode); ok {
+							valid = true
+							n := node.ParseCiliumNode(ciliumNode)
+							if n.IsLocal() {
+								equal = true
+								return
+							}
+							serNodes.Enqueue(func() error {
+								d.nodeDiscovery.Manager.NodeUpdated(n)
+								d.K8sEventProcessed(metricCiliumNode, metricUpdate, true)
+								return nil
+							}, serializer.NoRetry)
+						}
+					},
+					DeleteFunc: func(obj interface{}) {
+						var valid, equal bool
+						defer func() { d.K8sEventReceived(metricCiliumNode, metricDelete, valid, equal) }()
+						deletedObj, ok := obj.(cache.DeletedFinalStateUnknown)
```

Comment:
I'm not sure what is more common, @aanm. This is only hit on node delete which is a common operation anyway.

Reply:
The convention in this file is to add and use `k8s.CopyObjTo*` functions, and then try twice: https://github.com/cilium/cilium/blob/2e9bb515eb87439df4dd96936e7d3d92d07e0299/daemon/k8s_watcher.go#L1000-L1013

I did violate this recently, however, because it would create an import loop.

Reply:
I must have copied it from that bad example ;-)

Fixed.